File: sdks/java/src/main/java/com/featuresignals/sdk/FeatureSignalsProvider.java
package com.featuresignals.sdk;

import dev.openfeature.sdk.EvaluationContext;
import dev.openfeature.sdk.FeatureProvider;
import dev.openfeature.sdk.Metadata;
import dev.openfeature.sdk.ProviderEvaluation;
import dev.openfeature.sdk.ErrorCode;
import dev.openfeature.sdk.Reason;
import dev.openfeature.sdk.Structure;
import dev.openfeature.sdk.Value;
import dev.openfeature.sdk.Hook;

import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * OpenFeature-compliant provider backed by {@link FeatureSignalsClient}.
 *
 * <p>Implements the OpenFeature {@link FeatureProvider} interface. All evaluations
 * are local lookups against the client's cached flags &mdash; no network calls per
 * evaluation.
 *
 * <pre>{@code
 * FeatureSignalsProvider provider = new FeatureSignalsProvider("sdk-key", options);
 * OpenFeatureAPI.getInstance().setProviderAndWait(provider);
 * Client client = OpenFeatureAPI.getInstance().getClient();
 * boolean enabled = client.getBooleanValue("dark-mode", false);
 * }</pre>
 */
public class FeatureSignalsProvider implements FeatureProvider, AutoCloseable {

    private final FeatureSignalsClient client;

    public FeatureSignalsProvider(String sdkKey, ClientOptions options) {
        this.client = new FeatureSignalsClient(sdkKey, options);
    }

    public FeatureSignalsClient getClient() {
        return client;
    }

    @Override
    public Metadata getMetadata() {
        return () -> "FeatureSignals";
    }

    @Override
    public List<Hook> getProviderHooks() {
        return Collections.emptyList();
    }

    @Override
    public void initialize(EvaluationContext evaluationContext) throws Exception {
        client.waitForReady(30_000);
    }

    @Override
    public void shutdown() {
        client.close();
    }

    @Override
    public ProviderEvaluation<Boolean> getBooleanEvaluation(
            String flagKey, Boolean defaultValue, EvaluationContext ctx) {
        return resolve(flagKey, defaultValue, Boolean.class);
    }

    @Override
    public ProviderEvaluation<String> getStringEvaluation(
            String flagKey, String defaultValue, EvaluationContext ctx) {
        return resolve(flagKey, defaultValue, String.class);
    }

    @Override
    public ProviderEvaluation<Integer> getIntegerEvaluation(
            String flagKey, Integer defaultValue, EvaluationContext ctx) {
        Map<String, Object> flags = client.allFlags();
        Object val = flags.get(flagKey);
        if (val == null && !flags.containsKey(flagKey)) {
            return ProviderEvaluation.<Integer>builder()
                    .value(defaultValue)
                    .reason(Reason.ERROR.toString())
                    .errorCode(ErrorCode.FLAG_NOT_FOUND)
                    .errorMessage("flag '" + flagKey + "' not found")
                    .build();
        }
        if (val instanceof Number) {
            return ProviderEvaluation.<Integer>builder()
                    .value(((Number) val).intValue())
                    .reason(Reason.CACHED.toString())
                    .build();
        }
        return ProviderEvaluation.<Integer>builder()
                .value(defaultValue)
                .reason(Reason.ERROR.toString())
                .errorCode(ErrorCode.TYPE_MISMATCH)
                .errorMessage("expected numeric, got " + (val == null ? "null" : val.getClass().getSimpleName()))
                .build();
    }

    @Override
    public ProviderEvaluation<Double> getDoubleEvaluation(
            String flagKey, Double defaultValue, EvaluationContext ctx) {
        Map<String, Object> flags = client.allFlags();
        Object val = flags.get(flagKey);
        if (val == null && !flags.containsKey(flagKey)) {
            return ProviderEvaluation.<Double>builder()
                    .value(defaultValue)
                    .reason(Reason.ERROR.toString())
                    .errorCode(ErrorCode.FLAG_NOT_FOUND)
                    .errorMessage("flag '" + flagKey + "' not found")
                    .build();
        }
        if (val instanceof Number) {
            return ProviderEvaluation.<Double>builder()
                    .value(((Number) val).doubleValue())
                    .reason(Reason.CACHED.toString())
                    .build();
        }
        return ProviderEvaluation.<Double>builder()
                .value(defaultValue)
                .reason(Reason.ERROR.toString())
                .errorCode(ErrorCode.TYPE_MISMATCH)
                .errorMessage("expected numeric, got " + (val == null ? "null" : val.getClass().getSimpleName()))
                .build();
    }

    @Override
    public ProviderEvaluation<Value> getObjectEvaluation(
            String flagKey, Value defaultValue, EvaluationContext ctx) {
        Map<String, Object> flags = client.allFlags();
        Object val = flags.get(flagKey);
        if (val == null && !flags.containsKey(flagKey)) {
            return ProviderEvaluation.<Value>builder()
                    .value(defaultValue)
                    .reason(Reason.ERROR.toString())
                    .errorCode(ErrorCode.FLAG_NOT_FOUND)
                    .errorMessage("flag '" + flagKey + "' not found")
                    .build();
        }
        try {
            return ProviderEvaluation.<Value>builder()
                    .value(new Value(val))
                    .reason(Reason.CACHED.toString())
                    .build();
        } catch (InstantiationException e) {
            return ProviderEvaluation.<Value>builder()
                    .value(defaultValue)
                    .reason(Reason.ERROR.toString())
                    .errorCode(ErrorCode.TYPE_MISMATCH)
                    .errorMessage("unsupported value type: " +
                            (val == null ? "null" : val.getClass().getSimpleName()))
                    .build();
        }
    }

    @Override
    public void close() {
        shutdown();
    }

    @SuppressWarnings("unchecked")
    private <T> ProviderEvaluation<T> resolve(String flagKey, T defaultValue, Class<T> type) {
        Map<String, Object> flags = client.allFlags();
        Object val = flags.get(flagKey);
        if (val == null && !flags.containsKey(flagKey)) {
            return ProviderEvaluation.<T>builder()
                    .value(defaultValue)
                    .reason(Reason.ERROR.toString())
                    .errorCode(ErrorCode.FLAG_NOT_FOUND)
                    .errorMessage("flag '" + flagKey + "' not found")
                    .build();
        }
        if (type.isInstance(val)) {
            return ProviderEvaluation.<T>builder()
                    .value((T) val)
                    .reason(Reason.CACHED.toString())
                    .build();
        }
        return ProviderEvaluation.<T>builder()
                .value(defaultValue)
                .reason(Reason.ERROR.toString())
                .errorCode(ErrorCode.TYPE_MISMATCH)
                .errorMessage("expected " + type.getSimpleName() + ", got " +
                        (val == null ? "null" : val.getClass().getSimpleName()))
                .build();
    }
}

File: sdks/dotnet/src/FeatureSignals/OpenFeature/FeatureSignalsProvider.cs
using System.Text.Json;
using OpenFeature;
using OpenFeature.Model;

namespace FeatureSignals.OpenFeature;

/// <summary>
/// OpenFeature-compliant provider backed by <see cref="FeatureSignalsClient"/>.
///
/// All evaluations are local lookups against the client's cached flags.
///
/// <example>
/// <code>
/// var client = await FeatureSignalsClient.CreateAsync("sdk-key", new ClientOptions { EnvKey = "production" });
/// var provider = new FeatureSignalsProvider(client);
/// await Api.Instance.SetProviderAsync(provider);
/// var ofClient = Api.Instance.GetClient();
/// var enabled = await ofClient.GetBooleanValueAsync("dark-mode", false);
/// </code>
/// </example>
/// </summary>
public sealed class FeatureSignalsProvider : FeatureProvider, IDisposable
{
    private readonly FeatureSignalsClient _client;

    /// <summary>Creates a provider wrapping an existing client instance.</summary>
    public FeatureSignalsProvider(FeatureSignalsClient client)
    {
        _client = client;
    }

    /// <summary>Creates a provider that constructs its own client.</summary>
    public FeatureSignalsProvider(string sdkKey, ClientOptions options)
    {
        _client = new FeatureSignalsClient(sdkKey, options);
    }

    public FeatureSignalsClient Client => _client;

    public override Metadata GetMetadata() => new("FeatureSignals");

    public override async Task InitializeAsync(EvaluationContext context, CancellationToken cancellationToken = default)
    {
        await _client.WaitForReadyAsync(cancellationToken: cancellationToken);
    }

    public override Task ShutdownAsync(CancellationToken cancellationToken = default)
    {
        _client.Dispose();
        return Task.CompletedTask;
    }

    public override Task<ResolutionDetails<bool>> ResolveBooleanValueAsync(
        string flagKey, bool defaultValue, EvaluationContext? context = null, CancellationToken cancellationToken = default)
    {
        return Task.FromResult(Resolve(flagKey, defaultValue, val =>
            val is true or false ? (bool)val : throw new InvalidCastException()));
    }

    public override Task<ResolutionDetails<string>> ResolveStringValueAsync(
        string flagKey, string defaultValue, EvaluationContext? context = null, CancellationToken cancellationToken = default)
    {
        return Task.FromResult(Resolve(flagKey, defaultValue, val =>
            val is string s ? s : throw new InvalidCastException()));
    }

    public override Task<ResolutionDetails<int>> ResolveIntegerValueAsync(
        string flagKey, int defaultValue, EvaluationContext? context = null, CancellationToken cancellationToken = default)
    {
        return Task.FromResult(Resolve(flagKey, defaultValue, val => val switch
        {
            int i => i,
            long l => (int)l,
            double d => (int)d,
            _ => throw new InvalidCastException()
        }));
    }

    public override Task<ResolutionDetails<double>> ResolveDoubleValueAsync(
        string flagKey, double defaultValue, EvaluationContext? context = null, CancellationToken cancellationToken = default)
    {
        return Task.FromResult(Resolve(flagKey, defaultValue, val => val switch
        {
            double d => d,
            long l => l,
            int i => i,
            _ => throw new InvalidCastException()
        }));
    }

    public override Task<ResolutionDetails<Value>> ResolveStructureValueAsync(
        string flagKey, Value defaultValue, EvaluationContext? context = null, CancellationToken cancellationToken = default)
    {
        var flags = _client.AllFlags();
        if (!flags.TryGetValue(flagKey, out var raw))
        {
            return Task.FromResult(new ResolutionDetails<Value>(
                flagKey, defaultValue, reason: "ERROR", errorType: ErrorType.FlagNotFound,
                errorMessage: $"flag '{flagKey}' not found"));
        }

        try
        {
            var json = JsonSerializer.Serialize(raw);
            var structure = JsonSerializer.Deserialize<Value>(json) ?? defaultValue;
            return Task.FromResult(new ResolutionDetails<Value>(
                flagKey, structure, reason: "CACHED"));
        }
        catch
        {
            return Task.FromResult(new ResolutionDetails<Value>(
                flagKey, defaultValue, reason: "ERROR", errorType: ErrorType.TypeMismatch,
                errorMessage: $"cannot convert to Value"));
        }
    }

    public void Dispose() => _client.Dispose();

    private ResolutionDetails<T> Resolve<T>(
        string flagKey, T defaultValue, Func<object?, T> cast)
    {
        var flags = _client.AllFlags();
        if (!flags.TryGetValue(flagKey, out var raw))
        {
            return new ResolutionDetails<T>(
                flagKey, defaultValue, reason: "ERROR",
                errorType: ErrorType.FlagNotFound,
                errorMessage: $"flag '{flagKey}' not found");
        }

        try
        {
            return new ResolutionDetails<T>(
                flagKey, cast(raw), reason: "CACHED");
        }
        catch
        {
            return new ResolutionDetails<T>(
                flagKey, defaultValue, reason: "ERROR",
                errorType: ErrorType.TypeMismatch,
                errorMessage: $"expected {typeof(T).Name}, got {raw?.GetType().Name ?? "null"}");
        }
    }
}
